src/compiler/performance.ts

```diff
@@ -1,146 +1,209 @@
-        let enterCount = 0;
+    const nullTimer: Timer = { enter: noop, exit: noop };
+    export const performance = createPerformanceTracker();
+    export const buildPerformance = createPerformanceTracker();
+
+    function createPerformanceTracker() {
+        let perfHooks: PerformanceHooks | undefined;
+        // when set, indicates the implementation of `Performance` to use for user timing.
+        // when unset, indicates user timing is unavailable or disabled.
+        let performanceImpl: Performance | undefined;
+        let enabled = false;
+        let timeorigin = timestamp();
+        const marks = new Map<string, number>();
+        const counts = new Map<string, number>();
+        const durations = new Map<string, number>();
+        const durationMarks = new Set<string>();
+        let statistics: ESMap<string, Statistic> | undefined;
+
         return {
-            enter,
-            exit
+            createTimerIf,
+            createTimer,
+            mark,
+            measure,
+            addStatistics,
+            getCount,
+            getDuration,
+            forEachMeasure,
+            forEachCount,
+            forEachStatistics,
+            isEnabled,
+            enable,
+            disable,
         };
 
-        function enter() {
-            if (++enterCount === 1) {
-                mark(startMarkName);
+        function createTimerIf(condition: boolean, measureName: string, startMarkName: string, endMarkName: string) {
+            return condition ? createTimer(measureName, startMarkName, endMarkName) : nullTimer;
+        }
+
+        function createTimer(measureName: string, startMarkName: string, endMarkName: string): Timer {
+            let enterCount = 0;
+            return {
+                enter,
+                exit
+            };
+
+            function enter() {
+                if (++enterCount === 1) {
+                    mark(startMarkName);
+                }
+            }
+
+            function exit() {
+                if (--enterCount === 0) {
+                    mark(endMarkName);
+                    measure(measureName, startMarkName, endMarkName);
+                }
+                else if (enterCount < 0) {
+                    Debug.fail("enter/exit count does not match.");
+                }
             }
         }
 
-        function exit() {
-            if (--enterCount === 0) {
-                mark(endMarkName);
-                measure(measureName, startMarkName, endMarkName);
+        /**
+         * Marks a performance event.
+         *
+         * @param markName The name of the mark.
+         */
+        function mark(markName: string) {
+            if (enabled) {
+                const count = counts.get(markName) ?? 0;
+                counts.set(markName, count + 1);
+                marks.set(markName, timestamp());
+                performanceImpl?.mark(markName);

```

Comment:
This calls out to the native `PerformanceHooks` API for use with CPU profiles and a few other things in ts-perf. Is it intentional that both `performance` and `buildPerformance` call into this?